eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentRequest.scala

```diff
@@ -16,584 +16,41 @@
-    // README: since we use pubkey recovery to compute the node id from the message and signature, we don't check the signature.
-    // If instead we read the node id from the `n` field (which nobody uses afaict) then we would have to check the signature.
-    val amount_opt = Amount.decode(hrp.drop(prefix.length)) match {
-      case Success(value) => value
-      case Failure(e) => throw e
-    }
-    PaymentRequest(
-      prefix = prefix,
-      amount = amount_opt,
-      timestamp = bolt11Data.timestamp,
-      nodeId = pub,
-      tags = bolt11Data.taggedFields,
-      signature = bolt11Data.signature)
-  }
-
-  private def readBoltData(input: String): Bolt11Data = {
-    val lowercaseInput = input.toLowerCase
-    val separatorIndex = lowercaseInput.lastIndexOf('1')
-    val hrp = lowercaseInput.take(separatorIndex)
-    if (!prefixes.values.exists(prefix => hrp.startsWith(prefix))) throw new RuntimeException("unknown prefix")
-    val data = string2Bits(lowercaseInput.slice(separatorIndex + 1, lowercaseInput.length - 6)) // 6 == checksum size
-    Codecs.bolt11DataCodec.decode(data).require.value
-  }
-
-  /**
-   * Extracts the description from a serialized payment request that is **expected to be valid**.
-   * Throws an error if the payment request is not valid.
-   *
-   * @param input valid serialized payment request
-   * @return description as a String. If the description is a hash, returns the hash value as a String.
-   */
-  def fastReadDescription(input: String): String = {
-    readBoltData(input).taggedFields.collectFirst {
-      case PaymentRequest.Description(d) => d
-      case PaymentRequest.DescriptionHash(h) => h.toString()
-    }.get
-  }
-
-  /**
-   * Checks if a serialized payment request is expired. Timestamp is compared to the System's current time.
-   *
-   * @param input valid serialized payment request
-   * @return true if the payment request has expired, false otherwise.
-   */
-  def fastHasExpired(input: String): Boolean = {
-    val bolt11Data = readBoltData(input)
-    val expiry_opt = bolt11Data.taggedFields.collectFirst {
-      case p: PaymentRequest.Expiry => p
-    }
-    val timestamp = bolt11Data.timestamp
-    expiry_opt match {
-      case Some(expiry) => timestamp + expiry.toLong <= TimestampSecond.now()
-      case None => timestamp + DEFAULT_EXPIRY_SECONDS <= TimestampSecond.now()
-    }
-  }
-
-  /**
-   * @param pr payment request
-   * @return a bech32-encoded payment request
-   */
-  def write(pr: PaymentRequest): String = {
-    // currency unit is Satoshi, but we compute amounts in Millisatoshis
-    val hramount = Amount.encode(pr.amount)
-    val hrp = s"${pr.prefix}$hramount"
-    val data = Codecs.bolt11DataCodec.encode(Bolt11Data(pr.timestamp, pr.tags, pr.signature)).require
-    val int5s = eight2fiveCodec.decode(data).require.value
-    Bech32.encode(hrp, int5s.toArray)
+    Bolt11Invoice.read(input)
```

Comment:
Just to make sure I understand your plan: you kept this here instead of moving it to the `Bolt11Invoice` file because later you will check the human-readable prefix and automatically read it as a Bolt 11 or Bolt 12 invoice depending on the value?

Reply:
Exactly